File: fsrs-java/src/main/java/dev/bee/fsrs/FsrsEngine.java
package dev.bee.fsrs;

/**
 * Pure FSRS memory-math engine.
 */
public interface FsrsEngine {
    static FsrsEngine create(FsrsParameters parameters) {
        return new DefaultFsrsEngine(parameters);
    }

    static FsrsEngine latestDefault() {
        return create(FsrsParameters.latestDefault());
    }

    FsrsMemoryState initialState(FsrsRating firstRating);

    double retrievability(FsrsMemoryState state, int elapsedDays);

    default double nextDifficulty(double currentDifficulty, FsrsRating rating) {
        Fsrs.requireNonNull(rating, "rating");
        return nextState(new FsrsMemoryState(1.0, currentDifficulty), rating, 0).difficulty();
    }

    FsrsMemoryState nextState(FsrsMemoryState previousState, FsrsRating rating, int elapsedDays);

    double shortTermStability(double stability, FsrsRating rating);

    int nextIntervalDays(double stability, double desiredRetention, int maximumInterval);

    FsrsReviewOutput review(FsrsReviewInput input);
}

File: bee-fsrs/src/main/kotlin/dev/bee/fsrs/FsrsEngine.kt
package dev.bee.fsrs

/**
 * Pure FSRS memory-math engine.
 */
interface FsrsEngine {
    fun initialState(firstRating: FsrsRating?): FsrsMemoryState

    fun retrievability(state: FsrsMemoryState?, elapsedDays: Int): Double

    fun nextDifficulty(currentDifficulty: Double, rating: FsrsRating?): Double {
        val resolvedRating = Fsrs.requireNonNull(rating, "rating")
        return nextState(FsrsMemoryState(1.0, currentDifficulty), resolvedRating, 0).difficulty
    }

    fun nextState(previousState: FsrsMemoryState?, rating: FsrsRating?, elapsedDays: Int): FsrsMemoryState

    fun shortTermStability(stability: Double, rating: FsrsRating?): Double

    fun nextIntervalDays(stability: Double, desiredRetention: Double, maximumInterval: Int): Int

    fun review(input: FsrsReviewInput?): FsrsReviewOutput

    companion object {
        @JvmStatic
        fun create(parameters: FsrsParameters?): FsrsEngine = DefaultFsrsEngine(parameters)

        @JvmStatic
        fun latestDefault(): FsrsEngine = create(FsrsParameters.latestDefault())
    }
}
